### PR TITLE
Fixed #25617 -- Added case-insensitive unique username validation in UserCreationForm

### PR DESCRIPTION
**Pull request by pi-sigma**
Fixes [#25617](https://code.djangoproject.com/ticket/25617#comment:18)

Renamed original UserCreationForm to BaseUserCreationForm. The new UserCreationForm is a subclass of BaseUserCreationForm with one additional method that checks if a username already exists that differs only in case. Created one additional test ("test_user_already_iexists") for the new behavior and provided different username where necessary to make sure the old tests pass.

### CODE REVIEW

File: tests/auth_tests/test_forms.py [hunk 1]
```diff
@@ -213,9 +226,9 @@ class Meta(UserCreationForm.Meta):
                 fields = UserCreationForm.Meta.fields + ("date_of_birth",)
 
         data = {
-            "username": "testclient",
-            "password1": "testclient",
-            "password2": "testclient",
+            "username": "testclient2",
+            "password1": "testclient2",
+            "password2": "testclient2",
```
Comment (felixxm) [inline tests/auth_tests/test_forms.py:231]: This change looks unrelated :thinking: Please revert.

File: tests/auth_tests/test_forms.py [hunk 2]
```diff
@@ -191,9 +204,9 @@ def test_duplicate_normalized_unicode(self):
     )
     def test_validates_password(self):
         data = {
-            "username": "testclient",
-            "password1": "testclient",
-            "password2": "testclient",
+            "username": "testclient2",
+            "password1": "testclient2",
+            "password2": "testclient2",
```
Comment (felixxm) [inline tests/auth_tests/test_forms.py:209]: This change looks unrelated :thinking: Please revert.

File: django/contrib/auth/forms.py [hunk 3]
```diff
@@ -81,7 +81,7 @@ def widget_attrs(self, widget):
         }
 
 
-class UserCreationForm(forms.ModelForm):
+class BaseUserCreationForm(forms.ModelForm):
```
Comment (felixxm) [inline django/contrib/auth/forms.py:84]: We should document this class, add release notes, and add tests for a subclass of `BaseUserCreationForm`.

Should we add `BaseUserCreationForm.clean_username()`** stub?
```python
class BaseUserCreationForm(forms.ModelForm): 
    ...
    def clean_username(self):
        raise NotImplementedError(...)
```

File: tests/auth_tests/test_forms.py [hunk 4]
```diff
@@ -65,6 +65,19 @@ def test_user_already_exists(self):
...
+            "username": "Testclient",
+            "password1": "test123",
+            "password2": "test123",
+        }
+        form = UserCreationForm(data)
+        self.assertFalse(form.is_valid())
+        self.assertEqual(
+            form["username"].errors,
+            [str(User._meta.get_field("username").error_messages["unique"])],
+        )
```
Comment (felixxm) [inline tests/auth_tests/test_forms.py:79]: Additional test coverage for existing functionality should be merged in advance. Please move it to a separate commit.

**Review by felixxm — COMMENTED** [review]
@pi-sigma Thanks for this patch :+1:

File: tests/auth_tests/test_forms.py (hunk 2 above)
Comment (pi-sigma) [reply tests/auth_tests/test_forms.py]: I reverted the changes and renamed UserCreationFormTest to BaseUserCreationFormTest. The tests under the latter are now identical to the tests of the old UserCreationFormTest

File: django/contrib/auth/forms.py (hunk 3 above)
Comment (pi-sigma) [reply django/contrib/auth/forms.py:84]: Added test for UserCreationForm, release notes, and documentation for the new class. I might be missing out on test cases. So far I've only implemented the most obvious test for the new functionality.

Comment (pi-sigma) [reply django/contrib/auth/forms.py:84]: Adding this code to `BaseUserCreationForm` will result in a few of the tests erroring out. The `clean_username()` method depends on `getattr(self, "clean_%s" % name)()` in `~django/forms/forms.py` (line 452), which is used by calls to `is_valid()`. Hence every test that does `assertTrue(form.is_valid())` would produce the `NotImplementedError`.

**Comment by pi-sigma** [comment]
The docs fail to build due to a reference error. Not sure what I'm doing wrong. Am I not using the Sphinx syntax correctly? Thanks for the input!

File: docs/topics/auth/default.txt [hunk 5]
```diff
@@ -1665,6 +1665,14 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
     sets the user's password using
     :meth:`~django.contrib.auth.models.User.set_password()`.
 
+.. class:: UserCreationForm
+
+    Inherits from BaseUserCreationForm. Adds an additional method:
+
+    :meth:`~django.contrib.auth.forms.UserCreationForm.clean_username()`:
```
Comment (felixxm) [inline docs/topics/auth/default.txt:1672]: You cannot add a reference to sth that is not documented.

Comment (pi-sigma) [reply docs/topics/auth/default.txt:1672]: Makes sense. Will rebase and push again.

File: docs/releases/4.2.txt [hunk 6]
```diff
@@ -167,6 +167,13 @@ Forms
 * :class:`~django.forms.ModelForm` now accepts the new ``Meta`` option
   ``formfield_callback`` to customize form fields.
 
+* :class:`~django.contrib.auth.forms.BaseUserCreationForm` replaces the
+  old UserCreationForm.
+
+* :class:`~django.contrib.auth.forms.UserCreationForm` now inherits
+  from BaseUserCreationForm and adds a method to reject usernames
+  that differ only in case.
+
```
Comment (carltongibson) [inline docs/releases/4.2.txt:176]:
```suggestion
* The new :class:`~django.contrib.auth.forms.BaseUserCreationForm` is added,
  and is the recommended base class for customizing the user creation form.

* :class:`~django.contrib.auth.forms.UserCreationForm` now inherits from
  :class:`~django.contrib.auth.forms.BaseUserCreationForm`, and implements
  ``clean_username`` to reject usernames that differ only in case.
```

File: docs/topics/auth/default.txt [hunk 7]
```diff
@@ -1654,7 +1654,7 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
     A form used in the admin interface to change a user's information and
     permissions.
 
-.. class:: UserCreationForm
+.. class:: BaseUserCreationForm
```
Comment (carltongibson) [inline docs/topics/auth/default.txt:1657]: Needs a `versionadded` annotation, and should say ≈"This is the recommended base class if you need to customise the user creation form."

File: docs/topics/auth/default.txt [hunk 8]
```diff
@@ -1665,6 +1665,13 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
     sets the user's password using
     :meth:`~django.contrib.auth.models.User.set_password()`.
 
+.. class:: UserCreationForm
```
Comment (carltongibson) [inline docs/topics/auth/default.txt:1668]: `versionadded` annotation.

File: django/contrib/auth/forms.py [hunk 9]
```diff
@@ -144,6 +144,23 @@ def save(self, commit=True):
...
+    """
+    Adds a method for rejecting usernames that differ only in case. Does not
+    affect existing usernames or those created via createsuperuser.
+    """
+
+    def clean_username(self):
+        username = self.cleaned_data.get("username")
+        if username and User.objects.filter(username__iexact=username).exists():
+            raise forms.ValidationError(
+                User._meta.get_field("username").error_messages["unique"],
```
Comment (carltongibson) [inline django/contrib/auth/forms.py:157]: I think it would be better to declare an error message here on the form, so that when we're looking at it we don't need to go and track this down to know what it says.

File: tests/auth_tests/test_forms.py [hunk 10]
```diff
@@ -333,6 +334,21 @@ def test_html_autocomplete_attributes(self):
...
```
Comment (carltongibson) [inline tests/auth_tests/test_forms.py:349]: I might prefer the actual error message to the `User._meta.get_field("username").error_messages["unique"]` reference.

**Review by carltongibson — COMMENTED** [review]
Hi @pi-sigma. Thanks for this. Looking good!

I left a few comments.

**Comment by pi-sigma** [comment]
Thanks for the comments and suggestions! After rebasing on `upstream/main` I noticed that `UserCreationForm` had been modified by #16333 since I took up the ticket, hence there were some merge conflicts to deal with. Most of it is pretty straightforward, I think, but I'm not entirely sure about the documentation part. Let me know if this needs to be improved. Happy holidays!

File: django/contrib/auth/forms.py [hunk 11]
```diff
@@ -146,6 +146,20 @@ def save(self, commit=True):
...
+class UserCreationForm(BaseUserCreationForm):
+    """
+    Adds a method for rejecting usernames that differ only in case. Does not
+    affect existing usernames or those created via createsuperuser.
+    """
+
+    def clean_username(self):
+        username = self.cleaned_data.get("username")
+        if username and User.objects.filter(username__iexact=username).exists():
+            raise forms.ValidationError("A user with that username already exists.")
```
Comment (felixxm) [inline django/contrib/auth/forms.py:158]:
```suggestion
            raise forms.ValidationError(_("A user with that username already exists."), code="unique")
```

**Review by felixxm — COMMENTED** [review]
@pi-sigma Thanks :+1: I pushed final edits.

**Comment by felixxm** [comment]
@pi-sigma Welcome aboard :boat:
